File: coin-change/coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {

        vector<int> dp(amount+1,-1);
        sort(coins.begin(),coins.end());
        dp[0]=0;

        int i=1;
        while(i<=amount)
        {
            int change=INT_MAX;
            for(int j=0;j<coins.size();j++)
            {
                if(i<coins[j]) break;
                if(dp[i-coins[j]]==-1) continue;
                change=min(change,dp[i-coins[j]]+1);
            }
            if(change!=INT_MAX) dp[i]=change;
            i++;
        }
        return dp[amount];


        /*
        
        vector<int> v(amount+1,-1);

        sort(coins.begin(),coins.end());

        int n=coins.size();

        v[0]=0;

        for(int i=1;i<=amount;i++)
        {
            int j=0;
            int x=INT_MAX;
            while(j<n && i>=coins[j])
            {
                if(v[i-coins[j]]!=-1)
                {
                    x=min(x,v[i-coins[j]]+1);
                }
                j++;
            }
            if(x!=INT_MAX) v[i]=x;
        }

        return v[amount];

        */
    }
};

File: 322-coin-change/coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int n = coins.length;
        int[][] dp = new int[n+1][amount+1];
        for(int i=0;i<=n;i++) Arrays.fill(dp[i],Integer.MAX_VALUE);
        dp[0][0] = 0;
        for(int i=1;i<=n;i++)
        {
            for(int j=0;j<=amount;j++)
            {
                dp[i][j] = dp[i-1][j];
                if(coins[i-1]<=j && dp[i][j-coins[i-1]]!=Integer.MAX_VALUE)
                {
                    dp[i][j] = Math.min(dp[i][j],1+dp[i][j-coins[i-1]]);
                }
            }
        }

        return (dp[n][amount]==Integer.MAX_VALUE) ? -1 : dp[n][amount];

    }
}